<a href="https://colab.research.google.com/github/shahdarshil123/stay-or-stray-predict-course-dropouts/blob/dev/stay_or_stray.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Stay or Stray: Predict Course Dropout

Objective:
To tackle the challenge of predicting student dropout from Online Courses. Despite the growing popularity of online learning platforms, high dropout rates remain a significant issue, hindering the effectiveness of these educational resources. By accurately predicting the likelihood of a student dropping out from a course, the providers and instructors can implement targeted interventions and strategies to improve student engagement and retention.

Dataset: https://www.kaggle.com/competitions/stay-or-stray/data

### Downloading the dataset from Kaggle

In [ ]:
#Installing the required libraries
!pip install opendatasets --upgrade --quiet

In [ ]:
#Import all the required libraries
import opendatasets as od
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os


In [ ]:
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline

sns.set_style('darkgrid')
matplotlib.rcParams['font.size'] = 14
matplotlib.rcParams['figure.figsize'] = (9, 5)
matplotlib.rcParams['figure.facecolor'] = '#00000000'

In [ ]:
# Defining the dataset url
dataset_url = 'https://www.kaggle.com/competitions/stay-or-stray/data'

In [ ]:
# Downloading the dataset
od.download(dataset_url)

Skipping, found downloaded files in "./stay-or-stray" (use force=True to force download)


### Exploratory Dataset

In [ ]:
cwd = os.getcwd()
cwd

'/content'

In [ ]:
os.listdir('stay-or-stray')

['log_test.csv',
 'sample_submission.csv',
 'log_train.csv',
 'test.csv',
 'train.csv']

In [ ]:
train_path = cwd+'/stay-or-stray/train.csv'
train_path

'/content/stay-or-stray/train.csv'

In [ ]:
train_df = pd.read_csv(train_path)

In [ ]:
train_df.head(10)

ID        X1   X2    X3   X4   X5   X6   X7    X8   X9  ...  X131  \
0    7054.0  2.015152  0.0  32.0  NaN  4.0  NaN  NaN   7.0  6.0  ...   0.0   
1  147131.0  1.000000  NaN   2.0  NaN  NaN  0.0  NaN   6.0  5.0  ...   NaN   
2  132068.0  1.000000  NaN   1.0  NaN  2.0  NaN  0.0   9.0  1.0  ...   NaN   
3   85195.0  2.071429  0.0   1.0  NaN  2.0  NaN  NaN   8.0  7.0  ...   NaN   
4  191948.0  2.000000  NaN   2.0  0.0  2.0  NaN  NaN  14.0  2.0  ...   0.0   
5  196926.0  1.000000  NaN   1.0  NaN  2.0  NaN  0.0   6.0  3.0  ...   NaN   
6   88912.0  1.111111  NaN  15.0  NaN  1.0  NaN  NaN   9.0  1.0  ...   NaN   
7   39850.0  2.000000  NaN  72.0  0.0  1.0  NaN  NaN  63.0  4.0  ...   NaN   
8   37943.0  2.250000  NaN   2.0  0.0  2.0  0.0  NaN   8.0  1.0  ...   NaN   
9  178184.0  1.000000  NaN   2.0  NaN  1.0  0.0  NaN   6.0  6.0  ...   NaN   

   X132  X133  X134  X135  X136  X137  X138  X139  label  
0   NaN   NaN   1.0   1.0   1.0   3.0   2.0   NaN      1  
1   NaN   0.0   3.0   1.0   1.0   2.0  10.0   NaN      1  
2   NaN   0.0   1.0   1.0   2.0   2.0   4.0   NaN      1  
3   NaN   0.0   2.0   2.0   1.0   4.0   2.0   NaN      1  
4   0.0   NaN   2.0   2.0   1.0   2.0   6.0   0.0      1  
5   0.0   0.0   1.0   1.0   2.0   2.0   4.0   NaN      1  
6   0.0   NaN   6.0   1.0   2.0   1.0   3.0   NaN      1  
7   0.0   NaN   8.0   9.0   3.0   7.0   7.0   NaN      1  
8   NaN   NaN   1.0   4.0   1.0   2.0  10.0   0.0      1  
9   NaN   NaN   2.0   1.0   2.0   2.0   4.0   0.0      1  

[10 rows x 141 columns]

In [ ]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84377 entries, 0 to 84376
Columns: 141 entries, ID to label
dtypes: float64(140), int64(1)
memory usage: 90.8 MB


There are `141` columns in the training dataset and `84377` rows of data.

In [ ]:
train_df.describe()

ID            X1       X2            X3       X4  \
count   83912.000000  83959.000000  25179.0  83946.000000  25131.0   
mean   100386.093098      1.585512      0.0     11.740988      0.0   
std     57993.333045      0.535283      0.0     34.643444      0.0   
min         5.000000      1.000000      0.0      1.000000      0.0   
25%     50010.250000      1.000000      0.0      1.000000      0.0   
50%    100614.000000      2.000000      0.0      2.000000      0.0   
75%    150620.250000      2.000000      0.0      6.000000      0.0   
max    200906.000000      7.000000      0.0   3605.000000      0.0   

                 X5       X6       X7            X8            X9  ...  \
count  83947.000000  25137.0  25272.0  83967.000000  84008.000000  ...   
mean       1.906763      0.0      0.0      9.415139      9.026486  ...   
std        1.036347      0.0      0.0     17.288043     15.491942  ...   
min        1.000000      0.0      0.0      1.000000      1.000000  ...   
25%        1.000000      0.0      0.0      3.000000      3.000000  ...   
50%        2.000000      0.0      0.0      6.000000      6.000000  ...   
75%        2.000000      0.0      0.0      9.000000      9.000000  ...   
max       13.000000      0.0      0.0    487.000000    449.000000  ...   

          X131     X132     X133          X134          X135          X136  \
count  25084.0  25185.0  25367.0  83936.000000  83985.000000  83967.000000   
mean       0.0      0.0      0.0      2.381040      2.567792      2.254326   
std        0.0      0.0      0.0      2.050817      2.086498      1.989084   
min        0.0      0.0      0.0      1.000000      1.000000      1.000000   
25%        0.0      0.0      0.0      1.000000      1.000000      1.000000   
50%        0.0      0.0      0.0      2.000000      2.000000      2.000000   
75%        0.0      0.0      0.0      3.000000      3.000000      2.000000   
max        0.0      0.0      0.0     83.000000     87.000000     82.000000   

               X137          X138     X139         label  
count  83929.000000  83966.000000  24981.0  84377.000000  
mean       2.334092      5.735965      0.0      0.792728  
std        2.029105      2.988488      0.0      0.405355  
min        1.000000      1.000000      0.0      0.000000  
25%        1.000000      3.000000      0.0      1.000000  
50%        2.000000      6.000000      0.0      1.000000  
75%        2.000000      8.000000      0.0      1.000000  
max       73.000000     30.000000      0.0      1.000000  

[8 rows x 141 columns]

In [ ]:
# Check which columns contains null values
train_df.isnull().any()

ID        True
X1        True
X2        True
X3        True
X4        True
         ...  
X136      True
X137      True
X138      True
X139      True
label    False
Length: 141, dtype: bool

In [ ]:
# Find the missing ID from the dataset
missing_id = len(train_df[train_df['ID'].isnull()].index)
print(f'There are {missing_id} rows of missing IDs in the training set')

There are 465 rows of missing IDs in the training set


In [ ]:
total_count = train_df.label.count()
label_0 = train_df[train_df['label']==0].label.count()
label_1 = train_df[train_df['label']==1].label.count()
print(total_count)
print(label_0)
print(label_1)

84377
17489
66888


In [ ]:
frac_0 = label_0/total_count
frac_1 = label_1/total_count
print(frac_0)
print(frac_1)

0.20727212392002559
0.7927278760799744


In [ ]:
total_missing_count = missing_df.label.count()
missing_label_0 = missing_df[missing_df['label']==0].label.count()
missing_label_1 = missing_df[missing_df['label']==1].label.count()
print(total_missing_count)
print(missing_label_0)
print(missing_label_1)

465
97
368


In [ ]:
frac_missing_0 = missing_label_0/total_missing_count
frac_missing_1 = missing_label_1/total_missing_count
print(frac_missing_0)
print(frac_missing_1)

0.2086021505376344
0.7913978494623656


In [ ]:
#train_df.dropna(subset=['ID'],inplace=True)

We have dropped the 465 rows where the id is missing.

In [ ]:
#assert(train_df['ID'].isnull().sum()==0)

In [ ]:
total_count = train_df.label.count()
label_0 = train_df[train_df['label']==0].label.count()
label_1 = train_df[train_df['label']==1].label.count()
print(total_count)
print(label_0)
print(label_1)

84377
17489
66888


In [ ]:
frac_0 = label_0/total_count
frac_1 = label_1/total_count
print(frac_0)
print(frac_1)

0.20727212392002559
0.7927278760799744


Removing the samples with missing values will not affect the dataset.

As the prediction column i.e 'label' is highly skewed for values 1 and makes the dataset highly imbalanced we need to make it balance using methods like oversampling or undersampling.

### Create Input and Output/ Target dataframes

In [ ]:
input_df = train_df.drop(['label'],axis=1)
target_df = train_df['label']

In [ ]:
total_rows = input_df.shape[0]

In [ ]:
input_df.shape

(84377, 140)

In [ ]:
target_df.shape

(84377,)

In [ ]:
# X1 = train_df['X1']
# print(X1.dtype)
# print(X1.mean())
# plt.plot(X1)

In [ ]:
# fig, axes = plt.subplots(nrows=10, ncols=14)
# input_df.plot(subplots=True, ax=axes)
# plt.show()

### Data Cleaning and Preprocessing

We have to impute the missing values using various imputation methods


In [ ]:
#checking the columns for % of missing values
top_na_columns = (input_df.isna().sum()/total_count).sort_values(ascending=False).head(49)
top_na_columns.index

Index(['X139', 'X124', 'X51', 'X70', 'X91', 'X73', 'X58', 'X94', 'X131', 'X19',
       'X23', 'X61', 'X55', 'X53', 'X40', 'X4', 'X6', 'X63', 'X106', 'X60',
       'X59', 'X38', 'X2', 'X67', 'X132', 'X89', 'X20', 'X129', 'X119', 'X46',
       'X88', 'X24', 'X93', 'X18', 'X123', 'X75', 'X66', 'X7', 'X22', 'X54',
       'X21', 'X120', 'X71', 'X133', 'X76', 'X130', 'X62', 'X68', 'X39'],
      dtype='object')

In [ ]:
input_df[top_na_columns.index].describe()


X139     X124      X51      X70      X91      X73      X58      X94  \
count  24981.0  25044.0  25048.0  25055.0  25059.0  25060.0  25077.0  25077.0   
mean       0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
std        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
25%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
50%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
75%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
max        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

          X131      X19  ...      X54      X21     X120      X71     X133  \
count  25084.0  25093.0  ...  25311.0  25327.0  25337.0  25360.0  25367.0   
mean       0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
std        0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
min        0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
25%        0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
50%        0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
75%        0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   
max        0.0      0.0  ...      0.0      0.0      0.0      0.0      0.0   

           X76     X130      X62      X68      X39  
count  25385.0  25392.0  25418.0  25435.0  25436.0  
mean       0.0      0.0      0.0      0.0      0.0  
std        0.0      0.0      0.0      0.0      0.0  
min        0.0      0.0      0.0      0.0      0.0  
25%        0.0      0.0      0.0      0.0      0.0  
50%        0.0      0.0      0.0      0.0      0.0  
75%        0.0      0.0      0.0      0.0      0.0  
max        0.0      0.0      0.0      0.0      0.0  

[8 rows x 49 columns]

As the columns above are not describing any useful information about the data we can drop these columns from the data set

In [ ]:
input_df.drop(top_na_columns.index,axis=1,inplace=True)

In [ ]:
input_df.shape

(84377, 91)

In [ ]:
input_df

ID        X1    X3   X5    X8    X9   X10   X11   X12   X13  ...  \
0        7054.0  2.015152  32.0  4.0   7.0   6.0   2.0   5.0   2.0  15.0  ...   
1      147131.0  1.000000   2.0  NaN   6.0   5.0   8.0   4.0   7.0  12.0  ...   
2      132068.0  1.000000   1.0  2.0   9.0   1.0   7.0  10.0   7.0  17.0  ...   
3       85195.0  2.071429   1.0  2.0   8.0   7.0   9.0   4.0  18.0   8.0  ...   
4      191948.0  2.000000   2.0  2.0  14.0   2.0   2.0   4.0   4.0   3.0  ...   
...         ...       ...   ...  ...   ...   ...   ...   ...   ...   ...  ...   
84372   59741.0  2.000000   2.0  1.0   NaN   8.0   9.0   9.0   8.0  18.0  ...   
84373  126082.0  1.142857  40.0  1.0  10.0  42.0   2.0   1.0  19.0   7.0  ...   
84374   31267.0  2.333333  19.0  2.0   3.0   8.0   7.0   3.0   7.0  17.0  ...   
84375  132418.0  2.000000   1.0  NaN   2.0   8.0   5.0  10.0  10.0  15.0  ...   
84376   24271.0  2.000000   1.0  3.0   5.0   7.0  10.0   3.0  12.0  11.0  ...   

       X122  X125  X126  X127   X128  X134  X135  X136  X137  X138  
0      10.0   1.0   1.0  18.0   13.0   1.0   1.0   1.0   3.0   2.0  
1       1.0   1.0   1.0   9.0    4.0   3.0   1.0   1.0   2.0  10.0  
2       1.0   1.0   1.0   2.0    1.0   1.0   1.0   2.0   2.0   4.0  
3      14.0   1.0   1.0  12.0   12.0   2.0   2.0   1.0   4.0   2.0  
4       2.0   1.0   1.0  14.0    3.0   2.0   2.0   1.0   2.0   6.0  
...     ...   ...   ...   ...    ...   ...   ...   ...   ...   ...  
84372   2.0   1.0   1.0   5.0    1.0   2.0   2.0   3.0   1.0   3.0  
84373   2.0   1.0   1.0  12.0   22.0   4.0   4.0   2.0   1.0   6.0  
84374  10.0   1.0   1.0   9.0  168.0   2.0   2.0  17.0   5.0  11.0  
84375   1.0   1.0   1.0  15.0   10.0   3.0   2.0   1.0   2.0   7.0  
84376   2.0   1.0   1.0   2.0    1.0   1.0   1.0   2.0   2.0   7.0  

[84377 rows x 91 columns]

In [ ]:
input_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84377 entries, 0 to 84376
Data columns (total 91 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      83912 non-null  float64
 1   X1      83959 non-null  float64
 2   X3      83946 non-null  float64
 3   X5      83947 non-null  float64
 4   X8      83967 non-null  float64
 5   X9      84008 non-null  float64
 6   X10     83975 non-null  float64
 7   X11     83987 non-null  float64
 8   X12     83955 non-null  float64
 9   X13     83973 non-null  float64
 10  X14     83929 non-null  float64
 11  X15     83963 non-null  float64
 12  X16     83962 non-null  float64
 13  X17     83959 non-null  float64
 14  X25     83958 non-null  float64
 15  X26     83965 non-null  float64
 16  X27     83970 non-null  float64
 17  X28     83976 non-null  float64
 18  X29     83980 non-null  float64
 19  X30     83988 non-null  float64
 20  X31     83941 non-null  float64
 21  X32     83953 non-null  float64
 22

In [ ]:
input_df.describe()

ID            X1            X3            X5            X8  \
count   83912.000000  83959.000000  83946.000000  83947.000000  83967.000000   
mean   100386.093098      1.585512     11.740988      1.906763      9.415139   
std     57993.333045      0.535283     34.643444      1.036347     17.288043   
min         5.000000      1.000000      1.000000      1.000000      1.000000   
25%     50010.250000      1.000000      1.000000      1.000000      3.000000   
50%    100614.000000      2.000000      2.000000      2.000000      6.000000   
75%    150620.250000      2.000000      6.000000      2.000000      9.000000   
max    200906.000000      7.000000   3605.000000     13.000000    487.000000   

                 X9           X10           X11           X12           X13  \
count  84008.000000  83975.000000  83987.000000  83955.000000  83973.000000   
mean       9.026486      8.502042      7.922631     12.688667     12.633084   
std       15.491942     13.644927     12.402860     12.810191     12.963943   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        3.000000      3.000000      3.000000      6.000000      6.000000   
50%        6.000000      6.000000      6.000000     11.000000     11.000000   
75%        9.000000      9.000000      9.000000     17.000000     16.000000   
max      449.000000    370.000000    792.000000    417.000000    653.000000   

       ...          X122          X125          X126          X127  \
count  ...  83945.000000  83961.000000  83936.000000  83966.000000   
mean   ...      7.363285      1.142852      1.129956     11.095181   
std    ...     22.402294      2.659613      2.491213      8.140023   
min    ...      1.000000      1.000000      1.000000      1.000000   
25%    ...      1.000000      1.000000      1.000000      6.000000   
50%    ...      2.000000      1.000000      1.000000     11.000000   
75%    ...      3.000000      1.000000      1.000000     16.000000   
max    ...   1123.000000    239.000000    165.000000    531.000000   

               X128          X134          X135          X136          X137  \
count  83943.000000  83936.000000  83985.000000  83967.000000  83929.000000   
mean      20.785152      2.381040      2.567792      2.254326      2.334092   
std       38.831352      2.050817      2.086498      1.989084      2.029105   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        2.000000      1.000000      1.000000      1.000000      1.000000   
50%        6.000000      2.000000      2.000000      2.000000      2.000000   
75%       22.000000      3.000000      3.000000      2.000000      2.000000   
max     3642.000000     83.000000     87.000000     82.000000     73.000000   

               X138  
count  83966.000000  
mean       5.735965  
std        2.988488  
min        1.000000  
25%        3.000000  
50%        6.000000  
75%        8.000000  
max       30.000000  

[8 rows x 91 columns]

In [ ]:
input_df.isna().sum().sort_values()

X9      369
X110    370
X107    381
X105    386
X90     388
       ... 
X102    449
X116    453
X64     458
ID      465
X36     472
Length: 91, dtype: int64

In [ ]:
input_df.loc[:, input_df.isna().any()]

ID        X1    X3   X5    X8    X9   X10   X11   X12   X13  ...  \
0        7054.0  2.015152  32.0  4.0   7.0   6.0   2.0   5.0   2.0  15.0  ...   
1      147131.0  1.000000   2.0  NaN   6.0   5.0   8.0   4.0   7.0  12.0  ...   
2      132068.0  1.000000   1.0  2.0   9.0   1.0   7.0  10.0   7.0  17.0  ...   
3       85195.0  2.071429   1.0  2.0   8.0   7.0   9.0   4.0  18.0   8.0  ...   
4      191948.0  2.000000   2.0  2.0  14.0   2.0   2.0   4.0   4.0   3.0  ...   
...         ...       ...   ...  ...   ...   ...   ...   ...   ...   ...  ...   
84372   59741.0  2.000000   2.0  1.0   NaN   8.0   9.0   9.0   8.0  18.0  ...   
84373  126082.0  1.142857  40.0  1.0  10.0  42.0   2.0   1.0  19.0   7.0  ...   
84374   31267.0  2.333333  19.0  2.0   3.0   8.0   7.0   3.0   7.0  17.0  ...   
84375  132418.0  2.000000   1.0  NaN   2.0   8.0   5.0  10.0  10.0  15.0  ...   
84376   24271.0  2.000000   1.0  3.0   5.0   7.0  10.0   3.0  12.0  11.0  ...   

       X122  X125  X126  X127   X128  X134  X135  X136  X137  X138  
0      10.0   1.0   1.0  18.0   13.0   1.0   1.0   1.0   3.0   2.0  
1       1.0   1.0   1.0   9.0    4.0   3.0   1.0   1.0   2.0  10.0  
2       1.0   1.0   1.0   2.0    1.0   1.0   1.0   2.0   2.0   4.0  
3      14.0   1.0   1.0  12.0   12.0   2.0   2.0   1.0   4.0   2.0  
4       2.0   1.0   1.0  14.0    3.0   2.0   2.0   1.0   2.0   6.0  
...     ...   ...   ...   ...    ...   ...   ...   ...   ...   ...  
84372   2.0   1.0   1.0   5.0    1.0   2.0   2.0   3.0   1.0   3.0  
84373   2.0   1.0   1.0  12.0   22.0   4.0   4.0   2.0   1.0   6.0  
84374  10.0   1.0   1.0   9.0  168.0   2.0   2.0  17.0   5.0  11.0  
84375   1.0   1.0   1.0  15.0   10.0   3.0   2.0   1.0   2.0   7.0  
84376   2.0   1.0   1.0   2.0    1.0   1.0   1.0   2.0   2.0   7.0  

[84377 rows x 91 columns]

In [ ]:
input_df.std()

ID      57993.333045
X1          0.535283
X3         34.643444
X5          1.036347
X8         17.288043
            ...     
X134        2.050817
X135        2.086498
X136        1.989084
X137        2.029105
X138        2.988488
Length: 91, dtype: float64

In [ ]:
!pip install scikit-learn --upgrade --quiet

In [ ]:
from sklearn.impute import SimpleImputer

In [ ]:
imputer = SimpleImputer(strategy='mean')

In [ ]:
imput_cols = input_df.columns[input_df.columns != 'ID']
imput_cols

Index(['X1', 'X3', 'X5', 'X8', 'X9', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15',
       'X16', 'X17', 'X25', 'X26', 'X27', 'X28', 'X29', 'X30', 'X31', 'X32',
       'X33', 'X34', 'X35', 'X36', 'X37', 'X41', 'X42', 'X43', 'X44', 'X45',
       'X47', 'X48', 'X49', 'X50', 'X52', 'X56', 'X57', 'X64', 'X65', 'X69',
       'X72', 'X74', 'X77', 'X78', 'X79', 'X80', 'X81', 'X82', 'X83', 'X84',
       'X85', 'X86', 'X87', 'X90', 'X92', 'X95', 'X96', 'X97', 'X98', 'X99',
       'X100', 'X101', 'X102', 'X103', 'X104', 'X105', 'X107', 'X108', 'X109',
       'X110', 'X111', 'X112', 'X113', 'X114', 'X115', 'X116', 'X117', 'X118',
       'X121', 'X122', 'X125', 'X126', 'X127', 'X128', 'X134', 'X135', 'X136',
       'X137', 'X138'],
      dtype='object')

In [ ]:
imputer.fit(input_df[imput_cols])

SimpleImputer()

In [ ]:
print(list(imputer.statistics_))

[1.5855117845129942, 11.74098825435399, 1.9067626002120386, 9.415139280907976, 9.02648557280259, 8.502042274486454, 7.922630883350995, 12.688666547555238, 12.63308444380932, 12.55190696898569, 8.093160082417254, 8.052988256592268, 13.83142962636525, 12.464363133947927, 12.650914071339248, 12.71662498511373, 12.850088120415357, 12.793498452012384, 12.518324046292328, 12.41906815501364, 12.467690255261873, 12.566521355916054, 12.744222338701992, 12.7185519607376, 2.238424408557297, 10.132216347527866, 11.176522422539236, 10.173100753026404, 10.701429762897654, 2.238882934381327, 12.504930097411103, 11.031371380912665, 11.23663431191568, 27.01405414418599, 1.0009049127235492, 7.731510047527724, 7.316166195178879, 6.305293872446853, 1.0000271934994702, 1.0052115595368563, 1.0053785989410193, 10.774770696843358, 2480.7693068009958, 6.824397993226221, 9.021036832324773, 9.1823780989028, 8.278190025846564, 12.029473082283985, 9.518055787141803, 8.967389621371828, 8.999464062406956, 8.84953485

In [ ]:
print(list(np.round(imputer.statistics_,0)))

[2.0, 12.0, 2.0, 9.0, 9.0, 9.0, 8.0, 13.0, 13.0, 13.0, 8.0, 8.0, 14.0, 12.0, 13.0, 13.0, 13.0, 13.0, 13.0, 12.0, 12.0, 13.0, 13.0, 13.0, 2.0, 10.0, 11.0, 10.0, 11.0, 2.0, 13.0, 11.0, 11.0, 27.0, 1.0, 8.0, 7.0, 6.0, 1.0, 1.0, 1.0, 11.0, 2481.0, 7.0, 9.0, 9.0, 8.0, 12.0, 10.0, 9.0, 9.0, 9.0, 9.0, 10.0, 5.0, 9.0, 7.0, 9.0, 10.0, 11.0, 8.0, 8.0, 10.0, 12.0, 11.0, 11.0, 11.0, 12.0, 12.0, 12.0, 12.0, 13.0, 13.0, 12.0, 13.0, 4.0, 13.0, 13.0, 9.0, 5.0, 7.0, 1.0, 1.0, 11.0, 21.0, 2.0, 3.0, 2.0, 2.0, 6.0]


In [ ]:
print(list(imputer.statistics_))

[1.5855117845129942, 11.74098825435399, 1.9067626002120386, 9.415139280907976, 9.02648557280259, 8.502042274486454, 7.922630883350995, 12.688666547555238, 12.63308444380932, 12.55190696898569, 8.093160082417254, 8.052988256592268, 13.83142962636525, 12.464363133947927, 12.650914071339248, 12.71662498511373, 12.850088120415357, 12.793498452012384, 12.518324046292328, 12.41906815501364, 12.467690255261873, 12.566521355916054, 12.744222338701992, 12.7185519607376, 2.238424408557297, 10.132216347527866, 11.176522422539236, 10.173100753026404, 10.701429762897654, 2.238882934381327, 12.504930097411103, 11.031371380912665, 11.23663431191568, 27.01405414418599, 1.0009049127235492, 7.731510047527724, 7.316166195178879, 6.305293872446853, 1.0000271934994702, 1.0052115595368563, 1.0053785989410193, 10.774770696843358, 2480.7693068009958, 6.824397993226221, 9.021036832324773, 9.1823780989028, 8.278190025846564, 12.029473082283985, 9.518055787141803, 8.967389621371828, 8.999464062406956, 8.84953485

In [ ]:
imputer.statistics_ = np.round(imputer.statistics_,0)

In [ ]:
input_df[imput_cols] = imputer.transform(input_df[imput_cols])

In [ ]:
input_df.shape

(84377, 91)

In [ ]:
input_df = input_df.reset_index(drop=True)

In [ ]:
input_df

ID        X1    X3   X5    X8    X9   X10   X11   X12   X13  ...  \
0        7054.0  2.015152  32.0  4.0   7.0   6.0   2.0   5.0   2.0  15.0  ...   
1      147131.0  1.000000   2.0  2.0   6.0   5.0   8.0   4.0   7.0  12.0  ...   
2      132068.0  1.000000   1.0  2.0   9.0   1.0   7.0  10.0   7.0  17.0  ...   
3       85195.0  2.071429   1.0  2.0   8.0   7.0   9.0   4.0  18.0   8.0  ...   
4      191948.0  2.000000   2.0  2.0  14.0   2.0   2.0   4.0   4.0   3.0  ...   
...         ...       ...   ...  ...   ...   ...   ...   ...   ...   ...  ...   
84372   59741.0  2.000000   2.0  1.0   9.0   8.0   9.0   9.0   8.0  18.0  ...   
84373  126082.0  1.142857  40.0  1.0  10.0  42.0   2.0   1.0  19.0   7.0  ...   
84374   31267.0  2.333333  19.0  2.0   3.0   8.0   7.0   3.0   7.0  17.0  ...   
84375  132418.0  2.000000   1.0  2.0   2.0   8.0   5.0  10.0  10.0  15.0  ...   
84376   24271.0  2.000000   1.0  3.0   5.0   7.0  10.0   3.0  12.0  11.0  ...   

       X122  X125  X126  X127   X128  X134  X135  X136  X137  X138  
0      10.0   1.0   1.0  18.0   13.0   1.0   1.0   1.0   3.0   2.0  
1       1.0   1.0   1.0   9.0    4.0   3.0   1.0   1.0   2.0  10.0  
2       1.0   1.0   1.0   2.0    1.0   1.0   1.0   2.0   2.0   4.0  
3      14.0   1.0   1.0  12.0   12.0   2.0   2.0   1.0   4.0   2.0  
4       2.0   1.0   1.0  14.0    3.0   2.0   2.0   1.0   2.0   6.0  
...     ...   ...   ...   ...    ...   ...   ...   ...   ...   ...  
84372   2.0   1.0   1.0   5.0    1.0   2.0   2.0   3.0   1.0   3.0  
84373   2.0   1.0   1.0  12.0   22.0   4.0   4.0   2.0   1.0   6.0  
84374  10.0   1.0   1.0   9.0  168.0   2.0   2.0  17.0   5.0  11.0  
84375   1.0   1.0   1.0  15.0   10.0   3.0   2.0   1.0   2.0   7.0  
84376   2.0   1.0   1.0   2.0    1.0   1.0   1.0   2.0   2.0   7.0  

[84377 rows x 91 columns]

In [ ]:
input_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84377 entries, 0 to 84376
Data columns (total 91 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      83912 non-null  float64
 1   X1      84377 non-null  float64
 2   X3      84377 non-null  float64
 3   X5      84377 non-null  float64
 4   X8      84377 non-null  float64
 5   X9      84377 non-null  float64
 6   X10     84377 non-null  float64
 7   X11     84377 non-null  float64
 8   X12     84377 non-null  float64
 9   X13     84377 non-null  float64
 10  X14     84377 non-null  float64
 11  X15     84377 non-null  float64
 12  X16     84377 non-null  float64
 13  X17     84377 non-null  float64
 14  X25     84377 non-null  float64
 15  X26     84377 non-null  float64
 16  X27     84377 non-null  float64
 17  X28     84377 non-null  float64
 18  X29     84377 non-null  float64
 19  X30     84377 non-null  float64
 20  X31     84377 non-null  float64
 21  X32     84377 non-null  float64
 22

In [ ]:
input_df.describe()

ID            X1            X3            X5            X8  \
count   83912.000000  84377.000000  84377.000000  84377.000000  84377.000000   
mean   100386.093098      1.587565     11.742311      1.907238      9.413122   
std     57993.333045      0.534748     34.554855      1.033724     17.246013   
min         5.000000      1.000000      1.000000      1.000000      1.000000   
25%     50010.250000      1.000000      1.000000      1.000000      3.000000   
50%    100614.000000      2.000000      2.000000      2.000000      6.000000   
75%    150620.250000      2.000000      6.000000      2.000000      9.000000   
max    200906.000000      7.000000   3605.000000     13.000000    487.000000   

                X9           X10           X11           X12           X13  \
count  84377.00000  84377.000000  84377.000000  84377.000000  84377.000000   
mean       9.02637      8.504415      7.922988     12.690224     12.634841   
std       15.45803     13.612426     12.374164     12.778135     12.932895   
min        1.00000      1.000000      1.000000      1.000000      1.000000   
25%        3.00000      3.000000      3.000000      6.000000      6.000000   
50%        6.00000      6.000000      6.000000     11.000000     11.000000   
75%        9.00000      9.000000      9.000000     17.000000     16.000000   
max      449.00000    370.000000    792.000000    417.000000    653.000000   

       ...          X122          X125          X126          X127  \
count  ...  84377.000000  84377.000000  84377.000000  84377.000000   
mean   ...      7.361426      1.142148      1.129277     11.094718   
std    ...     22.344886      2.653068      2.484712      8.120177   
min    ...      1.000000      1.000000      1.000000      1.000000   
25%    ...      1.000000      1.000000      1.000000      6.000000   
50%    ...      2.000000      1.000000      1.000000     11.000000   
75%    ...      3.000000      1.000000      1.000000     16.000000   
max    ...   1123.000000    239.000000    165.000000    531.000000   

               X128          X134          X135          X136          X137  \
count  84377.000000  84377.000000  84377.000000  84377.000000  84377.000000   
mean      20.786257      2.379049      2.569800      2.253090      2.332318   
std       38.731359      2.045635      2.081853      1.984325      2.023857   
min        1.000000      1.000000      1.000000      1.000000      1.000000   
25%        2.000000      1.000000      1.000000      1.000000      1.000000   
50%        6.000000      2.000000      2.000000      2.000000      2.000000   
75%       21.000000      3.000000      3.000000      2.000000      2.000000   
max     3642.000000     83.000000     87.000000     82.000000     73.000000   

               X138  
count  84377.000000  
mean       5.737251  
std        2.981257  
min        1.000000  
25%        3.000000  
50%        6.000000  
75%        8.000000  
max       30.000000  

[8 rows x 91 columns]

In [ ]:
input_df.isna().sum()

ID      465
X1        0
X3        0
X5        0
X8        0
       ... 
X134      0
X135      0
X136      0
X137      0
X138      0
Length: 91, dtype: int64

### Spliting the dataset into train and validation dataset

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
train_input_df, val_input_df = train_test_split(input_df,test_size=0.25,random_state=42,shuffle=False)
train_label_df, val_label_df = train_test_split(target_df,test_size=0.25, random_state=42, shuffle=False)

In [ ]:
print(train_input_df.shape)
print(train_label_df.shape)

(63282, 91)
(63282,)


In [ ]:
train_input_df

ID        X1    X3   X5    X8   X9  X10   X11   X12   X13  ...  \
0        7054.0  2.015152  32.0  4.0   7.0  6.0  2.0   5.0   2.0  15.0  ...   
1      147131.0  1.000000   2.0  2.0   6.0  5.0  8.0   4.0   7.0  12.0  ...   
2      132068.0  1.000000   1.0  2.0   9.0  1.0  7.0  10.0   7.0  17.0  ...   
3       85195.0  2.071429   1.0  2.0   8.0  7.0  9.0   4.0  18.0   8.0  ...   
4      191948.0  2.000000   2.0  2.0  14.0  2.0  2.0   4.0   4.0   3.0  ...   
...         ...       ...   ...  ...   ...  ...  ...   ...   ...   ...  ...   
63277  116760.0  1.200000  11.0  2.0   7.0  7.0  6.0   1.0  16.0  20.0  ...   
63278  155994.0  2.000000   1.0  2.0   6.0  4.0  5.0   8.0  14.0  18.0  ...   
63279  113629.0  2.000000   2.0  1.0   7.0  4.0  6.0   7.0   7.0   4.0  ...   
63280  179935.0  1.000000  13.0  2.0   2.0  2.0  3.0   2.0   7.0   6.0  ...   
63281   42551.0  2.000000   1.0  1.0   2.0  3.0  5.0   5.0   4.0  10.0  ...   

       X122  X125  X126  X127  X128  X134  X135  X136  X137  X138  
0      10.0   1.0   1.0  18.0  13.0   1.0   1.0   1.0   3.0   2.0  
1       1.0   1.0   1.0   9.0   4.0   3.0   1.0   1.0   2.0  10.0  
2       1.0   1.0   1.0   2.0   1.0   1.0   1.0   2.0   2.0   4.0  
3      14.0   1.0   1.0  12.0  12.0   2.0   2.0   1.0   4.0   2.0  
4       2.0   1.0   1.0  14.0   3.0   2.0   2.0   1.0   2.0   6.0  
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
63277   1.0   6.0  25.0  12.0  56.0   4.0   2.0   6.0   5.0   5.0  
63278   1.0   1.0   1.0  17.0   9.0   2.0   2.0   1.0   1.0   6.0  
63279   1.0   1.0   1.0   4.0   1.0   2.0   1.0   1.0   1.0   3.0  
63280   1.0   1.0   1.0  11.0   4.0   1.0   1.0   3.0   1.0   5.0  
63281   2.0   1.0   1.0   4.0   1.0   1.0   1.0   1.0   3.0   6.0  

[63282 rows x 91 columns]

In [ ]:
train_label_df

0        1
1        1
2        1
3        1
4        1
        ..
63277    0
63278    1
63279    1
63280    1
63281    1
Name: label, Length: 63282, dtype: int64

In [ ]:
print(val_input_df.shape)
print(val_label_df.shape)

(21095, 91)
(21095,)


### Scaling the numeric columns

In [ ]:
from sklearn.preprocessing import MinMaxScaler

In [ ]:
scaler = MinMaxScaler()

In [ ]:
train_columns = train_input_df.columns[train_input_df.columns != 'ID']
train_columns

Index(['X1', 'X3', 'X5', 'X8', 'X9', 'X10', 'X11', 'X12', 'X13', 'X14', 'X15',
       'X16', 'X17', 'X25', 'X26', 'X27', 'X28', 'X29', 'X30', 'X31', 'X32',
       'X33', 'X34', 'X35', 'X36', 'X37', 'X41', 'X42', 'X43', 'X44', 'X45',
       'X47', 'X48', 'X49', 'X50', 'X52', 'X56', 'X57', 'X64', 'X65', 'X69',
       'X72', 'X74', 'X77', 'X78', 'X79', 'X80', 'X81', 'X82', 'X83', 'X84',
       'X85', 'X86', 'X87', 'X90', 'X92', 'X95', 'X96', 'X97', 'X98', 'X99',
       'X100', 'X101', 'X102', 'X103', 'X104', 'X105', 'X107', 'X108', 'X109',
       'X110', 'X111', 'X112', 'X113', 'X114', 'X115', 'X116', 'X117', 'X118',
       'X121', 'X122', 'X125', 'X126', 'X127', 'X128', 'X134', 'X135', 'X136',
       'X137', 'X138'],
      dtype='object')

In [ ]:
scaler.fit(train_input_df[train_columns])

MinMaxScaler()

In [ ]:
train_input_df[train_columns] = scaler.transform(train_input_df[train_columns])
val_input_df[train_columns] = scaler.transform(val_input_df[train_columns])

In [ ]:
train_input_df

ID        X1        X3        X5        X8        X9       X10  \
0        7054.0  0.169192  0.008602  0.250000  0.013605  0.011161  0.002890   
1      147131.0  0.000000  0.000277  0.083333  0.011338  0.008929  0.020231   
2      132068.0  0.000000  0.000000  0.083333  0.018141  0.000000  0.017341   
3       85195.0  0.178571  0.000000  0.083333  0.015873  0.013393  0.023121   
4      191948.0  0.166667  0.000277  0.083333  0.029478  0.002232  0.002890   
...         ...       ...       ...       ...       ...       ...       ...   
63277  116760.0  0.033333  0.002775  0.083333  0.013605  0.013393  0.014451   
63278  155994.0  0.166667  0.000000  0.083333  0.011338  0.006696  0.011561   
63279  113629.0  0.166667  0.000277  0.000000  0.013605  0.006696  0.014451   
63280  179935.0  0.000000  0.003330  0.083333  0.002268  0.002232  0.005780   
63281   42551.0  0.166667  0.000000  0.000000  0.002268  0.004464  0.011561   

            X11       X12       X13  ...      X122      X125      X126  \
0      0.005057  0.002404  0.021472  ...  0.008021  0.000000  0.000000   
1      0.003793  0.014423  0.016871  ...  0.000000  0.000000  0.000000   
2      0.011378  0.014423  0.024540  ...  0.000000  0.000000  0.000000   
3      0.003793  0.040865  0.010736  ...  0.011586  0.000000  0.000000   
4      0.003793  0.007212  0.003067  ...  0.000891  0.000000  0.000000   
...         ...       ...       ...  ...       ...       ...       ...   
63277  0.000000  0.036058  0.029141  ...  0.000000  0.021008  0.146341   
63278  0.008850  0.031250  0.026074  ...  0.000000  0.000000  0.000000   
63279  0.007585  0.014423  0.004601  ...  0.000000  0.000000  0.000000   
63280  0.001264  0.014423  0.007669  ...  0.000000  0.000000  0.000000   
63281  0.005057  0.007212  0.013804  ...  0.000891  0.000000  0.000000   

           X127      X128      X134      X135      X136      X137      X138  
0      0.032075  0.003296  0.000000  0.000000  0.000000  0.027778  0.034483  
1      0.015094  0.000824  0.024390  0.000000  0.000000  0.013889  0.310345  
2      0.001887  0.000000  0.000000  0.000000  0.012346  0.013889  0.103448  
3      0.020755  0.003021  0.012195  0.011628  0.000000  0.041667  0.034483  
4      0.024528  0.000549  0.012195  0.011628  0.000000  0.013889  0.172414  
...         ...       ...       ...       ...       ...       ...       ...  
63277  0.020755  0.015106  0.036585  0.011628  0.061728  0.055556  0.137931  
63278  0.030189  0.002197  0.012195  0.011628  0.000000  0.000000  0.172414  
63279  0.005660  0.000000  0.012195  0.000000  0.000000  0.000000  0.068966  
63280  0.018868  0.000824  0.000000  0.000000  0.024691  0.000000  0.137931  
63281  0.005660  0.000000  0.000000  0.000000  0.000000  0.027778  0.172414  

[63282 rows x 91 columns]

In [ ]:
train_input_df.describe()

ID            X1            X3            X5            X8  \
count   62922.000000  63282.000000  63282.000000  63282.000000  63282.000000   
mean   100232.789263      0.098062      0.002991      0.075977      0.019229   
std     57898.008009      0.089107      0.009967      0.086539      0.039504   
min         7.000000      0.000000      0.000000      0.000000      0.000000   
25%     50045.000000      0.000000      0.000000      0.000000      0.004535   
50%    100455.000000      0.166667      0.000277      0.083333      0.011338   
75%    150231.500000      0.166667      0.001387      0.083333      0.018141   
max    200906.000000      1.000000      1.000000      1.000000      1.000000   

                 X9           X10           X11           X12           X13  \
count  63282.000000  63282.000000  63282.000000  63282.000000  63282.000000   
mean       0.017912      0.021764      0.008722      0.028087      0.017826   
std        0.034288      0.039297      0.015611      0.030467      0.020018   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.004464      0.005780      0.002528      0.012019      0.007669   
50%        0.011161      0.014451      0.006321      0.024038      0.015337   
75%        0.017857      0.023121      0.010114      0.038462      0.023006   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

       ...          X122          X125          X126          X127  \
count  ...  63282.000000  63282.000000  63282.000000  63282.000000   
mean   ...      0.005729      0.000590      0.000788      0.019109   
std    ...      0.020314      0.011299      0.015346      0.015880   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.009434   
50%    ...      0.000891      0.000000      0.000000      0.018868   
75%    ...      0.001783      0.000000      0.000000      0.028302   
max    ...      1.000000      1.000000      1.000000      1.000000   

               X128          X134          X135          X136          X137  \
count  63282.000000  63282.000000  63282.000000  63282.000000  63282.000000   
mean       0.005447      0.016851      0.018286      0.015470      0.018582   
std        0.010798      0.025231      0.024336      0.024620      0.028356   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000275      0.000000      0.000000      0.000000      0.000000   
50%        0.001373      0.012195      0.011628      0.012346      0.013889   
75%        0.005768      0.024390      0.023256      0.012346      0.013889   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

               X138  
count  63282.000000  
mean       0.163468  
std        0.102970  
min        0.000000  
25%        0.068966  
50%        0.172414  
75%        0.241379  
max        1.000000  

[8 rows x 91 columns]

### Training the model using Logistic Regression


In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
model = LogisticRegression(solver='liblinear')

In [ ]:
model.fit(train_input_df[train_columns],train_label_df)

LogisticRegression(solver='liblinear')

In [ ]:
model.coef_

array([[-5.73929411e-02, -1.17605413e+00, -1.16963796e+00,
         1.20499838e+00,  6.20027228e-01, -2.09661237e-01,
         1.85479789e-01,  5.03756393e-01,  8.49106471e-01,
        -4.56644559e-03,  9.37550254e-01,  9.04584971e-01,
         2.57386741e-01, -5.36453328e-01,  2.97429781e-01,
        -1.49455250e-01,  8.30388220e-01,  1.24106797e+00,
         6.78539366e-01,  5.90599214e-02,  5.74023378e-02,
         8.37092689e-01,  7.78470959e-01, -5.16981595e+00,
        -2.63960965e-02,  2.04885250e+00, -9.09398234e-01,
        -1.67619129e+00, -2.14359658e+00, -3.49360636e-01,
        -1.94503416e+00, -8.66664581e-01, -1.40741704e+00,
        -6.42609027e-01,  2.10211954e-01,  3.17120753e+00,
        -2.30808821e+00,  2.02184548e+00, -3.68943687e-01,
         3.72638679e-01, -1.08924757e-01,  8.66299798e-02,
        -4.06525567e-01, -2.31567270e+00, -5.87828619e-01,
        -6.35007840e-01, -1.15370124e+00, -1.23881342e+00,
        -3.30074040e-01, -4.16928424e-01, -4.69041575e-0

In [ ]:
print(np.array(model.coef_.tolist()).shape)

(1, 90)


In [ ]:
print(model.intercept_)

[3.62604837]


In [ ]:
train_preds = model.predict(train_input_df[train_columns])
train_preds

array([1, 1, 1, ..., 1, 1, 1])

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
accuracy_score(train_label_df,train_preds)

0.8699788249423217

In [ ]:
val_preds = model.predict(val_input_df[train_columns])
val_preds

array([0, 1, 1, ..., 0, 1, 1])

In [ ]:
accuracy_score(val_label_df,val_preds)

0.86674567433041

### Model Predictions on Testing data

In [ ]:
# test data
test_path = cwd+'/stay-or-stray/test.csv'


In [ ]:
test_df = pd.read_csv(test_path)
test_df

ID        X1   X2    X3   X4   X5   X6   X7    X8    X9  ...  \
0       17547.0  1.272727  0.0   1.0  0.0  3.0  0.0  0.0   7.0  26.0  ...   
1      140449.0  1.000000  0.0   1.0  NaN  1.0  NaN  0.0   8.0   4.0  ...   
2      182658.0  2.000000  NaN   2.0  NaN  2.0  NaN  0.0  64.0   2.0  ...   
3      149652.0  1.000000  0.0   1.0  NaN  1.0  0.0  0.0   9.0   9.0  ...   
4      106304.0  2.000000  NaN   2.0  NaN  3.0  NaN  0.0   1.0   6.0  ...   
...         ...       ...  ...   ...  ...  ...  ...  ...   ...   ...  ...   
35977  102422.0  1.000000  NaN   2.0  NaN  1.0  NaN  NaN  10.0   7.0  ...   
35978  129387.0  2.076923  0.0   1.0  0.0  3.0  0.0  0.0   6.0   5.0  ...   
35979  131835.0  1.000000  NaN   1.0  0.0  1.0  NaN  NaN   7.0   9.0  ...   
35980   68715.0  1.500000  NaN  23.0  0.0  6.0  NaN  NaN   2.0  22.0  ...   
35981   55601.0  1.040000  NaN   1.0  NaN  2.0  NaN  NaN  10.0   6.0  ...   

       X130  X131  X132  X133  X134  X135  X136  X137  X138  X139  
0       NaN   NaN   NaN   0.0   2.0   6.0   2.0   4.0   3.0   NaN  
1       0.0   0.0   NaN   0.0   3.0   2.0   2.0   2.0   5.0   NaN  
2       NaN   NaN   NaN   NaN   2.0   4.0   1.0   1.0   7.0   0.0  
3       NaN   NaN   NaN   NaN   3.0   1.0   1.0   2.0   7.0   NaN  
4       NaN   NaN   NaN   0.0   1.0   2.0   2.0   2.0   2.0   0.0  
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
35977   NaN   0.0   NaN   0.0   3.0   1.0   1.0   2.0   3.0   NaN  
35978   NaN   NaN   NaN   0.0   2.0   2.0   2.0   2.0   2.0   0.0  
35979   NaN   NaN   NaN   NaN   3.0   1.0   2.0   1.0   7.0   NaN  
35980   0.0   0.0   0.0   0.0   1.0   4.0   1.0   1.0   9.0   NaN  
35981   NaN   NaN   NaN   0.0   2.0   2.0   3.0   1.0   6.0   0.0  

[35982 rows x 140 columns]

In [ ]:
# Drop the columns
test_df.drop(top_na_columns.index,axis=1,inplace=True)

In [ ]:
test_df.shape

(35982, 91)

In [ ]:
# Impute the missing data
test_df[imput_cols] = imputer.transform(test_df[imput_cols])

In [ ]:
test_df.isna().sum()

ID      0
X1      0
X3      0
X5      0
X8      0
       ..
X134    0
X135    0
X136    0
X137    0
X138    0
Length: 91, dtype: int64

In [ ]:
test_df

ID        X1    X3   X5    X8    X9   X10   X11   X12   X13  ...  \
0       17547.0  1.272727   1.0  3.0   7.0  26.0  48.0   8.0   8.0   9.0  ...   
1      140449.0  1.000000   1.0  1.0   8.0   4.0   5.0   3.0   1.0   6.0  ...   
2      182658.0  2.000000   2.0  2.0  64.0   2.0   7.0   8.0  14.0  19.0  ...   
3      149652.0  1.000000   1.0  1.0   9.0   9.0  10.0   1.0  15.0  19.0  ...   
4      106304.0  2.000000   2.0  3.0   1.0   6.0   4.0   2.0   4.0   4.0  ...   
...         ...       ...   ...  ...   ...   ...   ...   ...   ...   ...  ...   
35977  102422.0  1.000000   2.0  1.0  10.0   7.0   9.0  10.0   2.0  19.0  ...   
35978  129387.0  2.076923   1.0  3.0   6.0   5.0   9.0   7.0  13.0   4.0  ...   
35979  131835.0  1.000000   1.0  1.0   7.0   9.0   4.0   9.0   8.0  18.0  ...   
35980   68715.0  1.500000  23.0  6.0   2.0  22.0   7.0   8.0  17.0  19.0  ...   
35981   55601.0  1.040000   1.0  2.0  10.0   6.0   6.0   3.0   2.0   6.0  ...   

       X122  X125  X126  X127  X128  X134  X135  X136  X137  X138  
0      10.0   1.0   1.0  19.0  42.0   2.0   6.0   2.0   4.0   3.0  
1       1.0   1.0   1.0  17.0   3.0   3.0   2.0   2.0   2.0   5.0  
2      13.0   1.0   1.0  16.0  12.0   2.0   4.0   1.0   1.0   7.0  
3       2.0   1.0   1.0  16.0   3.0   3.0   1.0   1.0   2.0   7.0  
4       1.0   1.0   1.0   7.0   1.0   1.0   2.0   2.0   2.0   2.0  
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
35977   1.0   1.0   1.0   7.0   3.0   3.0   1.0   1.0   2.0   3.0  
35978   3.0   1.0   1.0   2.0  19.0   2.0   2.0   2.0   2.0   2.0  
35979   1.0   1.0   1.0   5.0   6.0   3.0   1.0   2.0   1.0   7.0  
35980   2.0   1.0   1.0  16.0   9.0   1.0   4.0   1.0   1.0   9.0  
35981  11.0   1.0   1.0   4.0  16.0   2.0   2.0   3.0   1.0   6.0  

[35982 rows x 91 columns]

In [ ]:
#Scale the testing data
test_df[train_columns] = scaler.transform(test_df[train_columns])

In [ ]:
test_df

ID        X1        X3        X5        X8        X9       X10  \
0       17547.0  0.045455  0.000000  0.166667  0.013605  0.055804  0.135838   
1      140449.0  0.000000  0.000000  0.000000  0.015873  0.006696  0.011561   
2      182658.0  0.166667  0.000277  0.083333  0.142857  0.002232  0.017341   
3      149652.0  0.000000  0.000000  0.000000  0.018141  0.017857  0.026012   
4      106304.0  0.166667  0.000277  0.166667  0.000000  0.011161  0.008671   
...         ...       ...       ...       ...       ...       ...       ...   
35977  102422.0  0.000000  0.000277  0.000000  0.020408  0.013393  0.023121   
35978  129387.0  0.179487  0.000000  0.166667  0.011338  0.008929  0.023121   
35979  131835.0  0.000000  0.000000  0.000000  0.013605  0.017857  0.008671   
35980   68715.0  0.083333  0.006104  0.416667  0.002268  0.046875  0.017341   
35981   55601.0  0.006667  0.000000  0.083333  0.020408  0.011161  0.014451   

            X11       X12       X13  ...      X122  X125  X126      X127  \
0      0.008850  0.016827  0.012270  ...  0.008021   0.0   0.0  0.033962   
1      0.002528  0.000000  0.007669  ...  0.000000   0.0   0.0  0.030189   
2      0.008850  0.031250  0.027607  ...  0.010695   0.0   0.0  0.028302   
3      0.000000  0.033654  0.027607  ...  0.000891   0.0   0.0  0.028302   
4      0.001264  0.007212  0.004601  ...  0.000000   0.0   0.0  0.011321   
...         ...       ...       ...  ...       ...   ...   ...       ...   
35977  0.011378  0.002404  0.027607  ...  0.000000   0.0   0.0  0.011321   
35978  0.007585  0.028846  0.004601  ...  0.001783   0.0   0.0  0.001887   
35979  0.010114  0.016827  0.026074  ...  0.000000   0.0   0.0  0.007547   
35980  0.008850  0.038462  0.027607  ...  0.000891   0.0   0.0  0.028302   
35981  0.002528  0.002404  0.007669  ...  0.008913   0.0   0.0  0.005660   

           X128      X134      X135      X136      X137      X138  
0      0.011261  0.012195  0.058140  0.012346  0.041667  0.068966  
1      0.000549  0.024390  0.011628  0.012346  0.013889  0.137931  
2      0.003021  0.012195  0.034884  0.000000  0.000000  0.206897  
3      0.000549  0.024390  0.000000  0.000000  0.013889  0.206897  
4      0.000000  0.000000  0.011628  0.012346  0.013889  0.034483  
...         ...       ...       ...       ...       ...       ...  
35977  0.000549  0.024390  0.000000  0.000000  0.013889  0.068966  
35978  0.004944  0.012195  0.011628  0.012346  0.013889  0.034483  
35979  0.001373  0.024390  0.000000  0.012346  0.000000  0.206897  
35980  0.002197  0.000000  0.034884  0.000000  0.000000  0.275862  
35981  0.004120  0.012195  0.011628  0.024691  0.000000  0.172414  

[35982 rows x 91 columns]

In [ ]:
test_df.describe()

ID            X1            X3            X5            X8  \
count   35982.000000  35982.000000  35982.000000  35982.000000  35982.000000   
mean   100392.184036      0.097857      0.002943      0.075790      0.019192   
std     58027.131420      0.089175      0.008835      0.085572      0.043336   
min         6.000000      0.000000      0.000000      0.000000      0.000000   
25%     50244.250000      0.000000      0.000000      0.000000      0.004535   
50%     99943.500000      0.166667      0.000277      0.083333      0.011338   
75%    150830.750000      0.166667      0.001387      0.083333      0.018141   
max    200902.000000      0.500000      0.264428      0.916667      3.367347   

                 X9           X10           X11           X12           X13  \
count  35982.000000  35982.000000  35982.000000  35982.000000  35982.000000   
mean       0.017462      0.021133      0.008641      0.027747      0.017766   
std        0.033398      0.037540      0.015188      0.030034      0.018742   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.004464      0.005780      0.002528      0.012019      0.007669   
50%        0.011161      0.014451      0.006321      0.024038      0.015337   
75%        0.017857      0.023121      0.010114      0.036058      0.023006   
max        1.071429      0.907514      0.694058      1.050481      0.526074   

       ...          X122          X125          X126          X127  \
count  ...  35982.000000  35982.000000  35982.000000  35982.000000   
mean   ...      0.005420      0.000540      0.000852      0.019036   
std    ...      0.018496      0.009952      0.017670      0.014269   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.009434   
50%    ...      0.000891      0.000000      0.000000      0.018868   
75%    ...      0.001783      0.000000      0.000000      0.028302   
max    ...      0.439394      0.714286      1.676829      0.364151   

               X128          X134          X135          X136          X137  \
count  35982.000000  35982.000000  35982.000000  35982.000000  35982.000000   
mean       0.005337      0.016767      0.017895      0.015539      0.018280   
std        0.011066      0.024870      0.023954      0.024350      0.027855   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000275      0.000000      0.000000      0.000000      0.000000   
50%        0.001373      0.012195      0.011628      0.012346      0.013889   
75%        0.005493      0.024390      0.023256      0.012346      0.013889   
max        0.963472      0.573171      0.767442      0.629630      0.597222   

               X138  
count  35982.000000  
mean       0.163371  
std        0.102761  
min        0.000000  
25%        0.068966  
50%        0.172414  
75%        0.241379  
max        0.793103  

[8 rows x 91 columns]

In [ ]:
test_pred = model.predict(test_df[train_columns])
test_pred

array([0, 1, 1, ..., 1, 1, 1])

In [ ]:
test_pred.shape

(35982,)

In [ ]:
values, count = np.unique(test_pred,return_counts=True)
print(values,count)

[0 1] [ 4468 31514]


### Creating submission file for Logistic Regression

In [ ]:
ID = test_df['ID'].values
ID

array([ 17547., 140449., 182658., ..., 131835.,  68715.,  55601.])

In [ ]:
submission_df = pd.DataFrame(data={'ID':ID, 'label':test_pred})
submission_df

ID  label
0       17547.0      0
1      140449.0      1
2      182658.0      1
3      149652.0      1
4      106304.0      1
...         ...    ...
35977  102422.0      1
35978  129387.0      1
35979  131835.0      1
35980   68715.0      1
35981   55601.0      1

[35982 rows x 2 columns]

In [ ]:
submission_df.to_csv('submission_logistic_regression.csv', index=None)

In [ ]:
from IPython.display import FileLink

In [ ]:
FileLink('submission.csv')

/content/submission.csv